document-portal/document-portal-fuse.c

```diff
@@ -3155,28 +3287,33 @@ xdp_fuse_init (GError **error)
       return FALSE;
     }
 
-  main_ch = fuse_mount (path, &args);
-  if (main_ch == NULL)
+  if (fuse_parse_cmdline (&args, &opts) != 0)
     {
       fuse_opt_free_args (&args);
-      g_set_error (error, XDG_DESKTOP_PORTAL_ERROR, XDG_DESKTOP_PORTAL_ERROR_FAILED,
-                   "Can't mount fuse fs on %s: %s", path, g_strerror (errno));
+      g_set_error (error, XDG_DESKTOP_PORTAL_ERROR,
+                   XDG_DESKTOP_PORTAL_ERROR_INVALID_ARGUMENT,
+                   "Impossible to parse command line");
       return FALSE;
     }
 
-  session = fuse_lowlevel_new (&args, &xdp_fuse_oper,
-                               sizeof (xdp_fuse_oper), NULL);
-  if (session == NULL)
+  thread_data.args = &args;
+  thread_data.loop_config.clone_fd = opts.clone_fd;
+  thread_data.loop_config.max_idle_threads = opts.max_idle_threads;
+
+  fuse_thread = g_thread_new ("fuse mainloop", xdp_fuse_thread, &thread_data);
+
+  while (g_atomic_pointer_get (&session) == NULL &&
+         g_atomic_pointer_get (&thread_data.error) == NULL)
+    ;
```

Comment:
Yeah, indeed this was the more classic pattern I was trying to avoid, as I tend to find it more complex than a manual handled one with atomic math, and seems to behave well. But indeed I may be wrong here.

I've changed the code though.

Thread sanitizer isn't too happy about this as it shows couple of races:

```
WARNING: ThreadSanitizer: data race (pid=173719)
  Write of size 8 at 0x55b2e5d00370 by thread T4:
    #0 xdp_fuse_thread document-portal/document-portal-fuse.c:3135 (xdg-document-portal+0x2dfb6)
    #1 g_thread_proxy ../../glib/glib/gthread.c:827 (libglib-2.0.so.0+0x80380)

  Previous read of size 8 at 0x55b2e5d00370 by main thread:
    #0 xdp_fuse_init document-portal/document-portal-fuse.c:3232 (xdg-document-portal+0x2e7e7)
    #1 on_name_acquired document-portal/document-portal.c:1460 (xdg-document-portal+0x1e1ce)
    #2 actually_do_call ../../glib/gio/gdbusnameowning.c:163 (libgio-2.0.so.0+0x121f13)
    #3 do_call ../../glib/gio/gdbusnameowning.c:215 (libgio-2.0.so.0+0x121f13)
    #4 __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58 (libc.so.6+0x2dfcf)

  Location is global 'session' of size 8 at 0x55b2e5d00370 (xdg-document-portal+0x000000059370)

  Thread T4 'fuse mainloop' (tid=173730, running) created by main thread at:
    #0 pthread_create ../../../../src/libsanitizer/tsan/tsan_interceptors_posix.cpp:969 (libtsan.so.0+0x605b8)
    #1 g_system_thread_new ../../glib/glib/gthread-posix.c:1323 (libglib-2.0.so.0+0xa9b14)
    #2 on_name_acquired document-portal/document-portal.c:1460 (xdg-document-portal+0x1e1ce)
    #3 actually_do_call ../../glib/gio/gdbusnameowning.c:163 (libgio-2.0.so.0+0x121f13)
    #4 do_call ../../glib/gio/gdbusnameowning.c:215 (libgio-2.0.so.0+0x121f13)
    #5 __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58 (libc.so.6+0x2dfcf)

SUMMARY: ThreadSanitizer: data race document-portal/document-portal-fuse.c:3135 in xdp_fuse_thread

----

WARNING: ThreadSanitizer: data race (pid=173719)
  Write of size 8 at 0x7b0400002110 by main thread:
    #0 free ../../../../src/libsanitizer/tsan/tsan_interceptors_posix.cpp:711 (libtsan.so.0+0x37ab8)
    #1 fuse_opt_free_args ../lib/fuse_opt.c:40 (libfuse3.so.3+0x1939c)
    #2 on_name_acquired document-portal/document-portal.c:1460 (xdg-document-portal+0x1e1ce)
    #3 actually_do_call ../../glib/gio/gdbusnameowning.c:163 (libgio-2.0.so.0+0x121f13)
    #4 do_call ../../glib/gio/gdbusnameowning.c:215 (libgio-2.0.so.0+0x121f13)
    #5 __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58 (libc.so.6+0x2dfcf)

  Previous write of size 8 at 0x7b0400002110 by thread T4:
    #0 malloc ../../../../src/libsanitizer/tsan/tsan_interceptors_posix.cpp:655 (libtsan.so.0+0x31c57)
    #1 __GI___strdup string/strdup.c:42 (libc.so.6+0xabbde)
    #2 fuse_opt_add_arg ../lib/fuse_opt.c:62 (libfuse3.so.3+0x1a2e2)
    #3 g_thread_proxy ../../glib/glib/gthread.c:827 (libglib-2.0.so.0+0x80380)

  Thread T4 'fuse mainloop' (tid=173730, running) created by main thread at:
    #0 pthread_create ../../../../src/libsanitizer/tsan/tsan_interceptors_posix.cpp:969 (libtsan.so.0+0x605b8)
    #1 g_system_thread_new ../../glib/glib/gthread-posix.c:1323 (libglib-2.0.so.0+0xa9b14)
    #2 on_name_acquired document-portal/document-portal.c:1460 (xdg-document-portal+0x1e1ce)
    #3 actually_do_call ../../glib/gio/gdbusnameowning.c:163 (libgio-2.0.so.0+0x121f13)
    #4 do_call ../../glib/gio/gdbusnameowning.c:215 (libgio-2.0.so.0+0x121f13)
    #5 __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58 (libc.so.6+0x2dfcf)

SUMMARY: ThreadSanitizer: data race ../lib/fuse_opt.c:40 in fuse_opt_free_args
```

And one more in case of error:

```
==================
WARNING: ThreadSanitizer: data race (pid=170251)
  Read of size 8 at 0x5617f6b5d380 by main thread:
    #0 xdp_fuse_exit document-portal/document-portal-fuse.c:3256 (xdg-document-portal+0x2e76c)
    #1 main document-portal/document-portal.c:1699 (xdg-document-portal+0x1ee65)

  Previous write of size 8 at 0x5617f6b5d380 by thread T4:
    #0 xdp_fuse_thread document-portal/document-portal-fuse.c:3108 (xdg-document-portal+0x2dd96)
    #1 g_thread_proxy ../../glib/glib/gthread.c:827 (libglib-2.0.so.0+0x80380)

  Location is global 'fuse_pthread' of size 8 at 0x5617f6b5d380 (xdg-document-portal+0x000000058380)

  Thread T4 'fuse mainloop' (tid=170821, finished) created by main thread at:
    #0 pthread_create ../../../../src/libsanitizer/tsan/tsan_interceptors_posix.cpp:969 (libtsan.so.0+0x605b8)
    #1 g_system_thread_new ../../glib/glib/gthread-posix.c:1323 (libglib-2.0.so.0+0xa9b14)
    #2 on_name_acquired document-portal/document-portal.c:1460 (xdg-document-portal+0x1e12e)
    #3 actually_do_call ../../glib/gio/gdbusnameowning.c:163 (libgio-2.0.so.0+0x121f13)
    #4 do_call ../../glib/gio/gdbusnameowning.c:215 (libgio-2.0.so.0+0x121f13)
    #5 __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58 (libc.so.6+0x2dfcf)

SUMMARY: ThreadSanitizer: data race document-portal/document-portal-fuse.c:3256 in xdp_fuse_exit
==================
ThreadSanitizer: reported 21 warnings
```

The one setting the session looks a false positive to me, as we're actually locking there, but the one on destruction may not and it's also one of the reason where using atomic math was giving some more control.

Reply:
I agree that multithreaded programming can be difficult. That's why I am suggesting that you use the primitives available: it makes it easier for others to reason about your code, and also means tools like ThreadSanitizer have a better idea of what your code is doing.

The current approach of "add atomic instructions until tsan stops complaining" is likely just hiding problems.